### PR TITLE
[Feature](datatype) Add IPv4/v6 data type for doris

### LINKED ISSUE #21370: [Feature](datatype) Add  IPv4/v6 data type for doris
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/doris/issues?q=is%3Aissue) and found no similar issues.

### Description

Now We do not support  IPv4/v6 data type for doris witch is common  data type in database
So it's the time to support it ! We have some basic funnctions for data type to check the data type is supported completly

- [ ] support CURD 
- [ ] support StreamLoad
- [ ] support Serde functions 
- [ ] support (implicit/explicit) cast functions 
- [ ] specific functions for itself data type (u can reference clickhouse or mysql)
        eg: IPv4ToIPv6(xxx) 

- [ ] make a Performance Test result with clickhouse or ElasticSearch in load this data type

### Use case

_No response_

### Related issues

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by sjyango**
## Proposed changes

Issue Number: close #21370

## Describe your changes
Now We do not support IPv4/v6 data type for doris witch is common data type in database
So it's the time to support it ! We have some basic funnctions for data type to check the data type is supported completly

- [x] support CURD

 - [x] support StreamLoad

 - [x] support Serde functions

 - [x] support (implicit/explicit) cast functions. The following cast functions have been implemented：
1. uint8/int8 -> ipv4
2. uint16/int16 -> ipv4
3. uint32/int32 -> ipv4
4. uint64/int64 -> ipv4
5. string -> ipv4
6. string -> ipv6

- [ ] specific functions for itself data type (u can reference clickhouse or mysql)

 - [ ] make a Performance Test result with clickhouse or ElasticSearch in load this data type

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by sjyango** [comment]
run buildall

**Comment by sjyango** [comment]
run p0

**Review by amorynan — APPROVED** [review]
LGTM

**Comment by amorynan** [comment]
@xiaokang  @yiguolei  @morrySnow please review this

**Comment by morrySnow** [comment]
please add rule descriptions for conversions with other types

File: fe/fe-common/src/main/java/org/apache/doris/catalog/Type.java [hunk 1]
```diff
@@ -1898,6 +1981,18 @@ public static Type getCmpType(Type t1, Type t2) {
                 || t2ResultType == PrimitiveType.LARGEINT)) {
             return Type.LARGEINT;
         }
+        if ((t1ResultType == PrimitiveType.IPV4
```
Comment (starocean999) [inline fe/fe-common/src/main/java/org/apache/doris/catalog/Type.java:1984]: should we handle IPV4 comp IPV6?

**Comment by sjyango** [comment]
> please add rule descriptions for conversions with other types

What is the format of rule descriptions for conversions? Are there any relevant examples?

**Comment by starocean999** [comment]
In general, there are more work need to do for nereids(at least):
1. Add literal definition of IPV4 and IPV6 like old planner
2. modify fe/fe-core/src/main/java/org/apache/doris/nereids/types/DataType.java for IP types
3. modify fe/fe-core/src/main/java/org/apache/doris/nereids/util/TypeCoercionUtils.java for IP types
4. add 'set enable_nereids_planner=true' in groovy regression test cases and verify the result

**Review by starocean999 — APPROVED** [review]

**Review by starocean999 — CHANGES_REQUESTED** [review]
In general, there are more work need to do for nereids(at least):

    Add literal definition of IPV4 and IPV6 like old planner
    modify fe/fe-core/src/main/java/org/apache/doris/nereids/types/DataType.java for IP types
    modify fe/fe-core/src/main/java/org/apache/doris/nereids/util/TypeCoercionUtils.java for IP types
    add 'set enable_nereids_planner=true' and 'enable_fallback_to_original_planner=false' in groovy regression test cases and verify the result

**Review by starocean999 — CHANGES_REQUESTED** [review]
In general, there are more work need to do for nereids(at least):

    Add literal definition of IPV4 and IPV6 like old planner
    modify fe/fe-core/src/main/java/org/apache/doris/nereids/types/DataType.java for IP types
    modify fe/fe-core/src/main/java/org/apache/doris/nereids/util/TypeCoercionUtils.java for IP types
    add 'set enable_nereids_planner=true' and 'enable_fallback_to_original_planner=false' in groovy regression test cases and verify the result

**Comment by sjyango** [comment]
> In general, there are more work need to do for nereids(at least):
> 
> 1. Add literal definition of IPV4 and IPV6 like old planner
> 2. modify fe/fe-core/src/main/java/org/apache/doris/nereids/types/DataType.java for IP types
> 3. modify fe/fe-core/src/main/java/org/apache/doris/nereids/util/TypeCoercionUtils.java for IP types
> 4. add 'set enable_nereids_planner=true' and 'enable_fallback_to_original_planner=false' in groovy regression test cases and verify the result

Thank you for reviewing my code, I will complete this work during the National Day!

**Review by starocean999 — CHANGES_REQUESTED** [review]
In general, there are more work need to do for nereids(at least):

    Add literal definition of IPV4 and IPV6 like old planner
    modify fe/fe-core/src/main/java/org/apache/doris/nereids/types/DataType.java for IP types
    modify fe/fe-core/src/main/java/org/apache/doris/nereids/util/TypeCoercionUtils.java for IP types
    add 'set enable_nereids_planner=true' and 'enable_fallback_to_original_planner=false' in groovy regression test cases and verify the result

**Comment by sjyango** [comment]
run p0

File: regression-test/suites/datatype_p0/ip/test_ip_crud.groovy [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+suite("test_ip_crud") {
+    sql """ DROP TABLE IF EXISTS test_unique_ip_crud """
+    sql """ DROP TABLE IF EXISTS test_dup_ip_crud """
+
+    sql """ SET enable_nereids_planner=true """
+    sql """ SET enable_fallback_to_original_planner=false """
+
+    sql """
+        CREATE TABLE test_unique_ip_crud (
```
Comment (starocean999) [inline regression-test/suites/datatype_p0/ip/test_ip_crud.groovy:27]: for dml statement, even 'SET enable_fallback_to_original_planner=false', the nereids will fallback to old planner if meets any error. So to support ipv4 and ipv6 keyword in nereids, we need modiy the g4 grammer files:
src/main/antlr4/org/apache/doris/nereids/DorisLexer.g4
src/main/antlr4/org/apache/doris/nereids/DorisParser.g4
and check visitCreateTable in org/apache/doris/nereids/parser/LogicalPlanBuilder.java works

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/literal/IPv6Literal.java
Comment (starocean999) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/literal/IPv6Literal.java:1]: checkValueValid is missing in nereids, shoud it be consistent with old planner?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/literal/IPv4Literal.java
Comment (starocean999) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/literal/IPv4Literal.java:1]: checkValueValid is missing in nereids, shoud it be consistent with old planner?

**Comment by sjyango** [comment]
run buildall

**Comment by sjyango** [comment]
run buildall

**Comment by sjyango** [comment]
run p0

File: fe/fe-core/src/main/java/org/apache/doris/analysis/BinaryPredicate.java [hunk 3]
```diff
@@ -411,6 +420,19 @@ private Type getCmpType() throws AnalysisException {
             }
         }
 
+        if (canCompareIP(getChild(0).getType().getPrimitiveType(), getChild(1).getType().getPrimitiveType())) {
+            if ((getChild(0).getType().isIP() && getChild(1) instanceof StringLiteral)
+                    || (getChild(1).getType().isIP() && getChild(0) instanceof StringLiteral)
+                    || (getChild(0).getType().isIP() && getChild(1).getType().isIP())) {
+                if (getChild(0).getType().isIPv4() || getChild(1).getType().isIPv4()) {
```
Comment (starocean999) [inline fe/fe-core/src/main/java/org/apache/doris/analysis/BinaryPredicate.java:427]: if IPV4 cmp IPV6, the getCmpType will return IPV4, is this the expected behavior?

**Comment by sjyango** [comment]
run buildall

**Comment by sjyango** [comment]
run feut

**Comment by sjyango** [comment]
run buildall

**Comment by sjyango** [comment]
run beut

**Comment by sjyango** [comment]
run beut

**Comment by sjyango** [comment]
run beut

**Comment by sjyango** [comment]
run p0

**Comment by sjyango** [comment]
run buildall

**Comment by sjyango** [comment]
run buildall

**Comment by sjyango** [comment]
run p0

**Review by starocean999 — DISMISSED** [review]
the FE part LGTM

**Comment by sjyango** [comment]
run buildall

**Comment by sjyango** [comment]
run buildall

**Comment by sjyango** [comment]
run buildall

**Comment by sjyango** [comment]
run buildall

**Comment by sjyango** [comment]
run buildall

**Comment by sjyango** [comment]
run p0

**Comment by sjyango** [comment]
run buildall

**Comment by sjyango** [comment]
run buildall

**Comment by sjyango** [comment]
run p0

**Comment by sjyango** [comment]
run External Regression

**Comment by sjyango** [comment]
run ext

**Comment by sjyango** [comment]
run buildall

**Comment by sjyango** [comment]
run buildall

**Comment by sjyango** [comment]
run buildall

**Comment by sjyango** [comment]
run clickbench-new

**Comment by sjyango** [comment]
run p0

**Comment by sjyango** [comment]
run p1

**Comment by sjyango** [comment]
run p0

**Review by starocean999 — APPROVED** [review]

**Pull request merged**